apps/prairielearn/src/lib/question-render.ts

```diff
@@ -468,89 +494,98 @@ export async function getAndRenderVariant(
-    locals.showTrueAnswer = true;
+    newLocals.showTrueAnswer = true;
   }
+  Object.assign(locals, newLocals);
 
   // We only fully render a small number of submissions on initial page
   // load; the rest only require basic information like timestamps. As
   // such, we'll load submissions in two passes: we'll load basic
   // information for all submissions to this variant, and then we'll
   // load the full submission for only the submissions that we'll
   // actually render.
-  const submissions = await sqldb.queryRows(
+  const basicSubmissions = await sqldb.queryRows(
     sql.select_basic_submissions,
-    { variant_id: locals.variant.id },
+    { variant_id: variant.id },
     SubmissionBasicSchema,
   );
-  const submissionCount = submissions.length;
+  const submissionCount = basicSubmissions.length;
+
+  const { submissions, submission } = await run(async () => {
+    if (submissionCount === 0) {
+      return { submissions: [], submission: null };
+    }
 
-  if (submissionCount >= 1) {
     // Load detailed information for the submissions that we'll render.
     // Note that for non-Freeform questions, we unfortunately have to
     // eagerly load detailed data for all submissions, as that ends up
     // being serialized in the HTML. v2 questions don't have any easy
     // way to support async rendering of submissions.
     const needsAllSubmissions = locals.question.type !== 'Freeform';
     const submissionsToRender = needsAllSubmissions
-      ? submissions
-      : submissions.slice(0, MAX_RECENT_SUBMISSIONS);
+      ? basicSubmissions
+      : basicSubmissions.slice(0, MAX_RECENT_SUBMISSIONS);
     const submissionDetails = await sqldb.queryRows(
       sql.select_detailed_submissions,
       { submission_ids: submissionsToRender.map((s) => s.id) },
       SubmissionDetailedSchema,
     );
 
-    locals.submissions = submissions.map((s, idx) => ({
+    const submissions = basicSubmissions.map((s, idx) => ({
       submission_number: submissionCount - idx,
       ...s,
       // Both queries order results consistently, so we can just use
       // the array index to match up the basic and detailed results.
       ...(idx < submissionDetails.length ? submissionDetails[idx] : {}),
     })) satisfies SubmissionForRender[];
-    locals.submission = locals.submissions[0]; // most recent submission
+    const submission = submissions[0]; // most recent submission
 
-    locals.showSubmissions = true;
     if (!locals.assessment && locals.question.show_correct_answer) {
       // instructor question pages, only show if true answer is
       // allowed by this question
-      locals.showTrueAnswer = true;
+      resultLocals.showTrueAnswer = true;
     }
-  }
 
-  locals.effectiveQuestionType = questionServers.getEffectiveQuestionType(locals.question.type);
+    return { submissions, submission };
+  });
+
+  resultLocals.submissions = submissions;
+  resultLocals.submission = submission;
```

Comment:
This removes the need for a separate `submission` variable, which can simplify the run function above:

```suggestion
  resultLocals.submission = submissions[0] ?? null;
```

Reply:
The `submission` variable is used elsewhere in this function. I want a local variable that's strongly typed as well; writes to `resultLocals` are strictly for usage by the callers of this function.

Reply:
The main idea stands, you can make the code simpler if you just assign to `submission` the first entry of `submissions` if it exists instead of setting it in the `run()` call. This makes the run call simpler, since it can return a single value and remove some of the clutter inside it.